cmake/Dependencies.cmake

```diff
@@ -221,47 +221,75 @@ if(USE_QNNPACK)
-        "${CONFU_DEPENDENCIES_BINARY_DIR}/QNNPACK")
-      # We build static versions of QNNPACK and pthreadpool but link
-      # them into a shared library for Caffe2, so they need PIC.
-      set_property(TARGET qnnpack PROPERTY POSITION_INDEPENDENT_CODE ON)
-      set_property(TARGET pthreadpool PROPERTY POSITION_INDEPENDENT_CODE ON)
-      set_property(TARGET cpuinfo PROPERTY POSITION_INDEPENDENT_CODE ON)
-    endif()
+    if (BUILD_CAFFE2_MOBILE)
+      if (NOT DEFINED QNNPACK_SOURCE_DIR)
+        set(QNNPACK_SOURCE_DIR "${CAFFE2_THIRD_PARTY_ROOT}/QNNPACK" CACHE STRING "QNNPACK source directory")
+      endif()
+
+      if(NOT TARGET qnnpack)
+        set(QNNPACK_BUILD_TESTS OFF CACHE BOOL "")
+        set(QNNPACK_BUILD_BENCHMARKS OFF CACHE BOOL "")
+        set(QNNPACK_CUSTOM_THREADPOOL ON CACHE BOOL "")
+        set(QNNPACK_LIBRARY_TYPE "static" CACHE STRING "")
+        set(PTHREADPOOL_LIBRARY_TYPE "static" CACHE STRING "")
+        set(CPUINFO_LIBRARY_TYPE "static" CACHE STRING "")
+        set(CPUINFO_LOG_LEVEL "error" CACHE STRING "")
+        add_subdirectory(
+          "${QNNPACK_SOURCE_DIR}"
+          "${CONFU_DEPENDENCIES_BINARY_DIR}/QNNPACK")
+        # We build static versions of QNNPACK and pthreadpool but link
+        # them into a shared library for Caffe2, so they need PIC.
+        set_property(TARGET qnnpack PROPERTY POSITION_INDEPENDENT_CODE ON)
+        set_property(TARGET pthreadpool PROPERTY POSITION_INDEPENDENT_CODE ON)
+        set_property(TARGET cpuinfo PROPERTY POSITION_INDEPENDENT_CODE ON)
+      endif()
+
+      list(APPEND Caffe2_DEPENDENCY_LIBS qnnpack)
+    else()
+      if (NOT DEFINED PYTORCH_QNNPACK_SOURCE_DIR)
+        set(PYTORCH_QNNPACK_SOURCE_DIR "${PROJECT_SOURCE_DIR}/aten/src/ATen/native/quantized/cpu/qnnpack" CACHE STRING "QNNPACK source directory")
+      endif()
 
-    list(APPEND Caffe2_DEPENDENCY_LIBS qnnpack)
+      if(NOT TARGET pytorch_qnnpack)
+        set(PYTORCH_QNNPACK_BUILD_TESTS OFF CACHE BOOL "")
+        set(PYTORCH_QNNPACK_BUILD_BENCHMARKS OFF CACHE BOOL "")
+        set(PYTORCH_QNNPACK_CUSTOM_THREADPOOL ON CACHE BOOL "")
+        set(PYTORCH_QNNPACK_LIBRARY_TYPE "static" CACHE STRING "")
+        set(PTHREADPOOL_LIBRARY_TYPE "static" CACHE STRING "")
+        set(CPUINFO_LIBRARY_TYPE "static" CACHE STRING "")
+        set(CPUINFO_LOG_LEVEL "error" CACHE STRING "")
+        add_subdirectory(
+          "${PYTORCH_QNNPACK_SOURCE_DIR}"
+          "${CONFU_DEPENDENCIES_BINARY_DIR}/pytorch_qnnpack")
+        # We build static versions of QNNPACK and pthreadpool but link
+        # them into a shared library for Caffe2, so they need PIC.
+        set_property(TARGET pytorch_qnnpack PROPERTY POSITION_INDEPENDENT_CODE ON)
+        set_property(TARGET pthreadpool PROPERTY POSITION_INDEPENDENT_CODE ON)
+        set_property(TARGET cpuinfo PROPERTY POSITION_INDEPENDENT_CODE ON)
+      endif()
+
+      set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -DUSE_PYTORCH_QNNPACK")
```

Comment:
I am a little unclear on the distinction between USE_QNNPACK and USE_PYTORCH_QNNPACK. Looks like for pytorch mobile they will both be set? But for caffe2 only USE_QNNPACK is set.

Reply:
USE_QNNPACK is the only user facing flag the user of the library needs to worry about at build time to avoid having two flags exposed at that level.  This flag is used to tell whether the user wants to build PyTorch / Caffe 2 with QNNPACK or not.  Beyond that though, and internally, we need to somehow be able to distinguish and differentiate between the scenario where we need to build the third party QNNPACK (for Caffe 2) or the one in PyTorch, otherwise we will end up with two versions of the library compiled in.  This is not going to cause any linker errors, because the symbols are prefixed with pytorch_ in the PyTorch version, but that means an unnecessary increase in library size.

One way to differentiate between the above two use cases is to use the value of USE_QNNPACK && !BUILD_CAFFE2_MOBILE to be able to tell this is a PyTorch mobile build (and consequently use the forked version of QNNPACK), and use USE_QNNPACK && BUILD_CAFFE2_MOBILE to designate it as a Caffe 2 build (and build the third party version of QNNPACK instead.)  As an aside, if USE_QNNPACK is OFF to begin with, neither version will be used.

Now that brings us to PYTORCH_USE_QNNPACK, which is defined whenever (USE_QNNPACK && !BUILD_CAFFE2_MOBILE) is true to avoid having to check for the combined flag every time in the source file.  If this is not a PyTorch build, that preprocessor variable will not be defined.  Again, USE_QNNPACK alone will not cut it because there would be no way to differentiate 3 scenarios (No QNNPACK, QNNPACK for Caffe 2, and QNNPACK for PyTorch) with a variable that can only hold two values.  The other option is to use the combination of two flags USE_QNNPACK && !BUILD_CAFFE2_MOBILE at all places which I believe is more error prone.

Reply:
Do we also set this macro for CI builds and runs in the jenkins setup? It may be getting automatically set, but just making sure.